design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,190 @@
+    active: 3
+    # where to run the nfs ganesha server
+    placement:
+    #  nodeAffinity:
+    #    requiredDuringSchedulingIgnoredDuringExecution:
+    #      nodeSelectorTerms:
+    #      - matchExpressions:
+    #        - key: role
+    #          operator: In
+    #          values:
+    #          - ganesha-node
+    #  tolerations:
+    #  - key: ganesha-node
+    #    operator: Exists
+    #  podAffinity:
+    #  podAntiAffinity:
+    # The requests and limits set here allow the ganesha pod(s) to use half
+    # of one CPU core and 1 gigabyte of memory
+    resources:
+    #  limits:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+    #  requests:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+
+  # Export settings
+  exports:
+    # The pseudoroot path. This is where the export will appear in the NFSv4
+    # pseudoroot namespace.
+  - pseudoPath: /cephfs/share1
+    # The directory in the exported file system this export is rooted on
+    path: /share1
+    allowedClients:
+    - clients:
+      - 172.17.0.0/16
+      - node1
+      accessMode: ReadOnly
+      squash: none
+
+  # NFS client recovery storage settings
+  clientRecovery:
+    # RADOS pool where NFS client recovery data is stored.
+    pool: myfs-data0
+    # RADOS namespace where NFS client recovery data is stored.
+    namespace: ganesha-ns
+
+```
+
+When the  nfs-ganesha.yaml is created the following will happen:
+
+- Rook's Ceph operator sees the creation of the NFS-Ganesha CRD and
+  stores the exports specified in the CRD as ConfigMaps.
+
+- The operator creates as many [k8s Deployments] as the number of active
+  Ganesha servers mentioned in the CRD. Each deployment brings up a Ganesha
+  server pod, a replicaset of size 1.
+
+- The ganesha servers, each running in a separate pod, use identical ganesha
+  config (ganesha.conf) that refer to exports stored as ConfigMaps.
+
+- The operator creates a k8s service for each of the ganesha server pods
+  to allow each of the them to have a stable IP address.
+
+The ganesha server pods constitute an active-active high availability NFS
+server cluster. If one of the active Ganesha server pods goes down, k8s brings
+up a replacment ganesha server pod with the same configuration and IP address.
+The NFS server cluster can be scaled up or down by updating the
+number of the active Ganesha servers in the CRD (using `kubectl edit` or
+modifying the original CRD and running `kubectly apply -f <CRD yaml file>`).
+
+### Dynamically adding, updating, and  removing exports of running Ganesha servers
+
+To dynamically update the exports of a Ganesha cluster you will have to update
+its CRD. On updating the CRD, the operator  will update the exports stored in
+the ConfigMaps, and will send SIGHUP signal or DBus message to the Ganesha
+servers to refresh their exports. Currently, there are issues with using either
+of the IPCs, SIGHUP and DBus, to dynamically update exports of the Ganesha
```

Comment:
I believe we would need something inside the ganesha pod that watches the file (mounted from the configmap) and when there is an update it will send the signal to ganesha. The design makes sense, we can figure out the implementation later...